pyvrp/cpp/Measure.h

```diff
@@ -0,0 +1,118 @@
+    friend class Measure<MeasureType::COST>;
+    friend class Measure<MeasureType::CAPACITY>;
+
+    value_type value;
+
+public:
+    // Default construction initialises to 0.
+    Measure() = default;
+
+    // This constructor takes any arithmetic type (generally useful) and casts
+    // its value to value_type.
+    template <typename T,
+              std::enable_if_t<std::is_arithmetic_v<T>, bool> = true>
+    Measure(T const value) : value(static_cast<value_type>(value))
+    {
+    }
+
+    // Explicit conversions of the underlying value to other arithmetic types.
+    template <typename T,
+              std::enable_if_t<std::is_arithmetic_v<T>, bool> = true>
+    explicit operator T() const
+    {
+        return static_cast<T>(value);
+    }
+
+    Measure &operator+=(Measure const &rhs)
+    {
+        this->value += rhs.value;
+        return *this;
+    }
+
+    Measure &operator-=(Measure const &rhs)
+    {
+        this->value -= rhs.value;
+        return *this;
+    }
+
+    Measure &operator*=(Measure const &rhs)
+    {
+        this->value *= rhs.value;
+        return *this;
+    }
+
+    // Explicit conversions to other measures.
+    explicit operator Measure<MeasureType::DISTANCE>() const { return {value}; }
+    explicit operator Measure<MeasureType::DURATION>() const { return {value}; }
+    explicit operator Measure<MeasureType::COST>() const { return {value}; }
+    explicit operator Measure<MeasureType::CAPACITY>() const { return {value}; }
+
+    // Comparison operators.
+    auto operator==(Measure const &other) const { return value == other.value; }
+    auto operator<=>(Measure const &other) const
+    {
+        return value <=> other.value;
+    }
+};
+
+template <MeasureType Type>
+Measure<Type> operator+(Measure<Type> const lhs, Measure<Type> const rhs)
+{
+    return static_cast<value_type>(lhs) + static_cast<value_type>(rhs);
+}
+
+template <MeasureType Type> Measure<Type> operator+(Measure<Type> const lhs)
+{
+    return +static_cast<value_type>(lhs);
+}
+
+template <MeasureType Type>
+Measure<Type> operator-(Measure<Type> const lhs, Measure<Type> const rhs)
+{
+    return static_cast<value_type>(lhs) - static_cast<value_type>(rhs);
+}
+
+template <MeasureType Type> Measure<Type> operator-(Measure<Type> const lhs)
+{
+    return -static_cast<value_type>(lhs);
+}
+
+template <MeasureType Type>
+Measure<Type> operator*(Measure<Type> const lhs, Measure<Type> const rhs)
+{
+    return static_cast<value_type>(lhs) * static_cast<value_type>(rhs);
+}
+
+template <MeasureType Type>
+Measure<Type> operator/(Measure<Type> const lhs, Measure<Type> const rhs)
+{
+    return static_cast<value_type>(lhs) / static_cast<value_type>(rhs);
+}
+
+// Useful type aliases.
+using distance_type = Measure<MeasureType::DISTANCE>;
+using duration_type = Measure<MeasureType::DURATION>;
+using cost_type = Measure<MeasureType::COST>;
+using capacity_type = Measure<MeasureType::CAPACITY>;
```

Comment:
Question: out of curiosity, why are the type aliases named as `cost_type` instead of `Cost`?

Reply:
That's a lot more consistent, so let's do that!